goguen/formal-spec/utxo.tex

```diff
@@ -542,117 +450,17 @@ \subsection{Two-Phase Transaction Validation for Non-Native Scripts}
 result in different validation outcomes running the same script on the same
 arguments.
 
-
 \begin{figure}[htb]
-  \begin{equation}
-    \inference[Scripts-Val]
-    {
-    \fun{txvaltag}~\var{tx} = \True  &
-    \var{remExU}~\geq~0
-    }
-    {
-    \begin{array}{l}
-      \var{pp}\\
-      \var{tx}\\
-    \end{array}
-      \vdash
-      \left(
-      \begin{array}{r}
-        \var{remExU}\\
-      \end{array}
-      \right)
-      \trans{sval}{\epsilon}
-      \left(
-      \begin{array}{r}
-        \var{remExU}\\
-      \end{array}
-      \right) \\
-    }
-  \end{equation}
-  \begin{equation}
-    \inference[Scripts-Stop]
-    { \\~\\
-    (\var{isVal},\var{remExU'})~:=~ \llbracket sc \rrbracket_
-    {cm,\var{remExU}} dt \\
-    (sc, dt, cm) := s
-    \\
-    ~
-    \\
-    \fun{txvaltag}~\var{tx} = \False &
-    (\var{remExU'}~<~0 ~ \lor ~ \var{isVal} = \False)
-    }
-    {
-    \begin{array}{l}
-      \var{pp}\\
-      \var{tx}\\
-    \end{array}
-      \vdash
-      \left(
-      \begin{array}{r}
-        \var{remExU}\\
-      \end{array}
-      \right)
-      \trans{sval}{\Gamma;s}
-      \left(
-      \begin{array}{r}
-        \var{remExU}\\
-      \end{array}
-      \right)
-    }
-  \end{equation}
-  \begin{equation}
-    \inference[Scripts-Ind]
-    {
-    {
-    \begin{array}{l}
-      \var{pp}\\
-      \var{tx}\\
-    \end{array}
-    }
-      \vdash
-      \left(
-      {
-      \begin{array}{r}
-        \var{remExU}\\
-      \end{array}
-      }
-      \right)
-      \trans{sval}{\Gamma}
-      \left(
-      {
-      \begin{array}{r}
-        \var{remExU'}\\
-      \end{array}
-      }
-      \right) \\
-    (\var{isVal},\var{remExU''})~:=~ \llbracket sc \rrbracket
-    _{cm,\var{remExU'}} dt \\
-    (sc, dt, cm) := s & \var{remExU''}~\geq~0
-    }
-    {
-    \begin{array}{l}
-      \var{pp}\\
-      \var{tx}\\
-    \end{array}
-      \vdash
-      \left(
-      \begin{array}{r}
-        \var{remExU}\\
-      \end{array}
-      \right)
-      \trans{sval}{\Gamma;s}
-      \left(
-      \begin{array}{r}
-        \varUpdate{remExU''}\\
-      \end{array}
-      \right)
-    }
-  \end{equation}
-  \caption{Script validation rules}
-  \label{fig:rules:utxo-scrval}
+  \begin{align*}
+    & \fun{evalScripts} \in \seqof{(\ScriptPurpose \times \Script \times \seqof{Data})} \to \ExUnits \to \Bool \\
+    & \fun{evalScripts}~\epsilon~\var{remExU}~=~\True \\
+    & \fun{evalScripts}~((\var{it}, \var{script}, \var{d});\Gamma)~\var{remExU}~=~
+      \var{isVal} \land \fun{evalScripts}~\Gamma~\var{remExU'} \\
+    & ~~\where \\
+    & ~~~~ (\var{isVal},\var{remExU'})~:=~ \llbracket sc \rrbracket_{cm,\var{remExU}} d
```

Comment:
we used to check that the execution units never went negative. are we now assuming that `isVal` will return False if ever the units dry up?

Reply:
Yeah, in fact this was also how we assumed it would behave in the previous version. I'm pretty sure that check was redundant, possibly a leftover from some older version.